src/rgw/rgw_data_sync.cc

```diff
@@ -4136,12 +4138,83 @@ static bool has_olh_epoch(RGWModifyOp op) {
 }
 
+class RGWBucketShardIsDoneCR: public RGWCoroutine {
+  RGWDataSyncCtx *sc;
+  RGWDataSyncEnv *sync_env;
+  rgw_bucket_sync_status bucket_status;
+  const rgw_raw_obj status_obj;
+  const int shard_id;
+  RGWObjVersionTracker objv_tracker;
+  const bilog_list_result& bilog_list;
+  const uint64_t generation;
+
+public:
+  RGWBucketShardIsDoneCR(RGWDataSyncCtx *_sc, const std::string& _status_oid,
+                         int _shard_id, bilog_list_result& _bilog_list, const uint64_t _gen)
+    : RGWCoroutine(_sc->cct), sc(_sc), sync_env(_sc->env),
+      status_obj(sync_env->svc->zone->get_zone_params().log_pool, _status_oid),
+      shard_id(_shard_id), bilog_list(_bilog_list), generation(_gen) {}
+
+  int operate() override
+  {
+    reenter(this) {
+      do {
+        // read bucket sync status
+        objv_tracker.clear();
+        using ReadCR = RGWSimpleRadosReadCR<rgw_bucket_sync_status>;
+        yield call(new ReadCR(sync_env->async_rados, sync_env->svc->sysobj,
+                              status_obj, &bucket_status, false, &objv_tracker));
+        if (retcode < 0) {
+          ldout(cct, 20) << "failed to read bucket shard status: "
+              << cpp_strerror(retcode) << dendl;
+          return set_cr_error(retcode);
+        }
+
+        // exit with success to avoid stale shard being
+        // retried in error repo if we lost a race
+        if (bucket_status.state != BucketSyncState::Incremental ||
+            bucket_status.incremental_gen != generation) {
+            return set_cr_done();
+            }
+
+        yield {
+          // update bucket_status after a shard is done with current gen
+          auto& done = bucket_status.shards_done_with_gen;
+          done[shard_id] = true;
+
+          // increment gen if all shards are already done with current gen
+          if (std::all_of(done.begin(), done.end(),
+            [] (const bool done){return done; } )) {
+            bucket_status.incremental_gen = bilog_list.next_log->generation;
+            done.clear();
+            done.resize(bilog_list.next_log->num_shards, false);
+          }
+          using WriteCR = RGWSimpleRadosWriteCR<rgw_bucket_sync_status>;
+          call(new WriteCR(sync_env->async_rados, sync_env->svc->sysobj,
+                            status_obj, bucket_status, &objv_tracker, false));
+          if (retcode < 0 && retcode != -ECANCELED) {
+            ldout(cct, 20) << "failed to write bucket sync status: " << cpp_strerror(retcode) << dendl;
+            return set_cr_error(retcode);
+          } else if (retcode >= 0) {
+            return set_cr_done();
+          }
+        }
+
+      } while (retcode == -ECANCELED);
+
+    }
+
+    return 0;
+  }
+};
+
 class RGWBucketShardIncrementalSyncCR : public RGWCoroutine {
   RGWDataSyncCtx *sc;
   RGWDataSyncEnv *sync_env;
   rgw_bucket_sync_pipe& sync_pipe;
   RGWBucketSyncFlowManager::pipe_rules_ref rules;
   rgw_bucket_shard& bs;
+  const std::string& status_oid;
```

Comment:
is this the bucket-wide status oid, or the shard status oid? we should name the variable accordingly. since it's coming `RGWBucketShardIncrementalSyncCR`, i assume we're accidentally using the shard oid here

Reply:
right, fixed it. thanks!